### PR TITLE
Fix compilation for 3ds Max 2015 & 2017

### PR DESCRIPTION
**Pull request by usakhelo**
Hi,

This PR should fix interactive rendering for max 2015 and max 2017. Most of the changes address max 2015 sdk. It has less support for interactive rendering and some of things has to be done manually. Pressing render button during ongoing rendering is one of those things - max 2016 and 2017 sdk-s handle cancellation and restart, while in 2015 I couldn't find suitable methods.

Regards,
Sergo.

### CODE REVIEW

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #include "appleseedinteractive/interactivesession.h"
 #include "appleseedinteractive/interactivetilecallback.h"
 #include "appleseedrenderer/projectbuilder.h"
+#include "appleseedrenderer/appleseedrenderer.h"
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Order alphabetically.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 2]
```diff
@@ -50,3 +55,14 @@ void InteractiveRendererController::set_status(const Status status)
 {
     m_status = status;
 }
+
+#if MAX_RELEASE == MAX_RELEASE_R17
+void InteractiveRendererController::on_progress()
+{
+    auto progress_cb = dynamic_cast<IRenderProgressCallback*>(m_progress_cb);
+
+    m_status = progress_cb->Progress(0, 0) == RENDPROG_CONTINUE
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: More conventionally:

    if (progress_cb->Progress(0, 0) == RENDPROG_CONTINUE)
        m_status = AbortRendering;

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 3]
```diff
@@ -31,17 +31,30 @@
...
+
+// Forward declarations.
+class RendProgressCallback;
+
 class InteractiveRendererController
   : public renderer::DefaultRendererController
 {
   public:
-    InteractiveRendererController();
+    InteractiveRendererController(RendProgressCallback* progress_cb);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Mark the constructor `explicit` to prevent implicit consutrction from a `RendProgressCallback`.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp [hunk 4]
```diff
@@ -71,6 +76,10 @@ void InteractiveSession::render_thread()
 
     // Render the frame.
     renderer->render();
+
+#if MAX_RELEASE == MAX_RELEASE_R17
+    static_cast<AppleseedInteractiveRender*>(m_renderer)->AbortRender();
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp]: It's a bit weird to call `AbortRender()` at the end of the render. Do you really need to do anything? Wouldn't calling `EndSession()` be more logical?

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 5]
```diff
@@ -40,12 +40,16 @@ namespace renderer { class Project; }
 class Bitmap;
 class IIRenderMgr;
 class InteractiveRendererController;
+class RendProgressCallback;
+class IInteractiveRender;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 6]
```diff
@@ -58,10 +62,12 @@ class InteractiveSession
   private:
     void render_thread();
 
+    RendProgressCallback*                           m_progress_cb;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: It is *very* important that the order of data member declarations and the order of data member initialization match. This doesn't appear to be the case right now.

File: src/appleseed-max-impl/appleseedinteractive/interactivetilecallback.cpp [hunk 7]
```diff
@@ -32,6 +32,7 @@
 // 3ds Max headers.
 #include <bitmap.h>
 #include <interactiverender.h>
+#include <max.h>
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivetilecallback.cpp]: Why is this header required?

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 8]
```diff
@@ -108,14 +109,27 @@ void AppleseedRenderer::DeleteThis()
...
 void* AppleseedRenderer::GetInterface(ULONG id)
 {
+#if MAX_RELEASE == MAX_RELEASE_R17 || MAX_RELEASE == MAX_RELEASE_R18
     if (id == I_RENDER_ID)
     {
+#if MAX_RELEASE == MAX_RELEASE_R17
+        if (g_interactive_renderer != nullptr)
+        {
+            g_interactive_renderer->EndSession();
+            g_interactive_renderer = nullptr;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: You're leaking memory here since you're not deleting `g_interactive_renderer`.

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.h [hunk 9]
```diff
@@ -128,6 +128,8 @@ class AppleseedRenderer
     virtual IOResult Save(ISave* isave) override;
     virtual IOResult Load(ILoad* iload) override;
 
+    static AppleseedInteractiveRender* g_interactive_renderer;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.h]: I'm not sure to understand why we need a static data member here (it's a global variable and _will_ cause problems in multithreading scenarios).

File: src/appleseed-max-impl/logtarget.cpp [hunk 10]
```diff
@@ -104,7 +104,11 @@ void LogTarget::write(
         break;
     }
 
+#if MAX_RELEASE > MAX_RELEASE_R17
     auto is_ui_thread = GetCOREInterface15()->GetMainThreadID() == GetCurrentThreadId();
+#else
+    auto is_ui_thread = reinterpret_cast<Interface15*>(GetCOREInterface()->GetInterface(Interface15::kInterface15InterfaceID))->GetMainThreadID() == GetCurrentThreadId();
```
Comment (dictoon) [inline src/appleseed-max-impl/logtarget.cpp]: Factorize the part that retrieves an `Interface` pointer.

File: src/appleseed-max-impl/appleseedinteractive/interactivetilecallback.cpp (hunk 7 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedinteractive/interactivetilecallback.cpp]: max sdk 2017 complains about undefined Interface type. For some reason 2016 sdk compiles without any error.
I'll replace max.h with maxapi.h, it seems to bring less headers into the project.

**Comment by usakhelo** [comment]
Most of the commented parts are gone with removing interactive mode from max 2015.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 11]
```diff
@@ -29,10 +29,13 @@
...
 #include "interactiverenderercontroller.h"
 
+// 3ds Max headers.
+#include <interactiverender.h>
+
 namespace asr = renderer;
 
 InteractiveRendererController::InteractiveRendererController()
-  : m_status(ContinueRendering)
+    : m_status(ContinueRendering)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: Revert this change.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.h [hunk 12]
```diff
@@ -89,7 +94,9 @@ class AppleseedInteractiveRender
     virtual BOOL IsRendering() override;
 
     // IAbortableRenderer methods.
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.h]: Move the comment inside the `#if` block.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp [hunk 13]
```diff
@@ -50,3 +53,4 @@ void InteractiveRendererController::set_status(const Status status)
 {
     m_status = status;
 }
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.cpp]: Revert change.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 14]
```diff
@@ -39,9 +48,9 @@ class InteractiveRendererController
 
     virtual void on_rendering_begin() override;
     virtual Status get_status() const override;
-    
+
     void set_status(const Status status);
 
   private:
-    Status m_status;
+    Status                  m_status;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Revert change.

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 15]
```diff
@@ -73,7 +73,6 @@ namespace
 
 AppleseedRendererClassDesc g_appleseed_renderer_classdesc;
 
-
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: Revert change.

File: src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp [hunk 16]
```diff
@@ -108,14 +107,18 @@ void AppleseedRenderer::DeleteThis()
 
 void* AppleseedRenderer::GetInterface(ULONG id)
 {
+#if MAX_RELEASE == MAX_RELEASE_R18  // No interactive render for max 2015.
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: What about Max 2017 and later?

Comment (usakhelo) [reply src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp]: 2017 and later use special methods instead of GetInterface.
